File: py_capi/pytest/sat/adder_test.py
#! /usr/bin/env python3

"""SatSolver.add_adder() に関するテスト

:file: adder_test.py
:author: Yusuke Matsunaga (松永 裕介)
:copyright: Copyright (C) 2023 Yusuke Matsunaga, All rights reserved.
"""

import pytest
from ymworks.sat import SatBool3, SatSolver
from solver_test import SolverTest


class AdderTest(SolverTest):

    def __init__(self, solver_type=None):
        super().__init__(solver_type)

    def check_adder(self, na, nb, ns):
        assert na <= ns
        assert nb <= ns
        nall = na + nb + ns + 2
        assert nall <= self.var_num

        alits = [ self.var_list[i] for i in range(na) ]
        blits = [ self.var_list[i + na] for i in range(nb) ]
        ilit = self.var_list[na + nb]
        slits = [ self.var_list[i + na + nb + 1] for i in range(ns) ]
        olit = self.var_list[na + nb + ns + 1]

        self.solver.add_adder(alits, blits, ilit, slits, olit)

        nexp = 1 << nall
        vals = [ None for _ in range(nexp) ]
        amask = (1 << na) - 1
        ashift = 0
        bmask = (1 << nb) - 1
        bshift = na
        imask = 1
        ishift = na + nb
        smask = (1 << ns) - 1
        sshift = na + nb + 1
        omask = 1
        oshift = na + nb + 1 + ns
        for p in range(nexp):
            a = (p >> ashift) & amask
            b = (p >> bshift) & bmask
            i = (p >> ishift) & imask
            s = (p >> sshift) & smask
            o = (p >> oshift) & omask
            s_exp = a + b + i
            o_exp = 0
            if s_exp >= (1 << ns):
                o_exp = 1
                s_exp -= (1 << ns)
            if s == s_exp and o == o_exp:
                vals[p] = 1
            else:
                vals[p] = 0
        self.check(nall, vals)

@pytest.fixture
def sat_test_fixture():
    return AdderTest()

def test_add_4_4_4(sat_test_fixture):
    sat_test_fixture.check_adder(4, 4, 4)

def test_add_4_3_4(sat_test_fixture):
    sat_test_fixture.check_adder(4, 3, 4)

def test_add_4_2_4(sat_test_fixture):
    sat_test_fixture.check_adder(4, 2, 4)

def test_add_3_4_4(sat_test_fixture):
    sat_test_fixture.check_adder(3, 4, 4)

def test_add_2_4_4(sat_test_fixture):
    sat_test_fixture.check_adder(2, 4, 4)

def test_add_3_3_4(sat_test_fixture):
    sat_test_fixture.check_adder(3, 3, 4)

def test_add_3_2_4(sat_test_fixture):
    sat_test_fixture.check_adder(3, 2, 4)

def test_add_3_1_4(sat_test_fixture):
    sat_test_fixture.check_adder(3, 1, 4)

def test_add_3_1_5(sat_test_fixture):
    sat_test_fixture.check_adder(3, 1, 5)

File: c++-srcs/sat/gtest/adder_test.cc

/// @file adder_test.cc
/// @brief SatSolver::add_adder() のテストプログラム
/// @author Yusuke Matsunaga (松永 裕介)
///
/// Copyright (C) 2025 Yusuke Matsunaga
/// All rights reserved.

#include "gtest/gtest.h"
#include "SatTestFixture.h"
#include "ym/Range.h"


BEGIN_NAMESPACE_YM

class AdderTest :
  public SatTestFixture
{
public:

  /// @brief コンストラクタ
  AdderTest() : SatTestFixture() { }

  /// @brief デストラクタ
  ~AdderTest() = default;


public:
  //////////////////////////////////////////////////////////////////////
  // 外部インターフェイス
  //////////////////////////////////////////////////////////////////////

  /// @brief adder のチェックを行う．
  void
  check(
    int na, ///< [in] a のビット幅
    int nb, ///< [in] b のビット幅
    int ns  ///< [in] s のビット幅
  );

};

// @brief adder のチェックを行う．
void
AdderTest::check(
  int na,
  int nb,
  int ns
)
{
  ASSERT_COND( na <= ns );
  ASSERT_COND( nb <= ns );

  ASSERT_COND( na + nb + ns + 2 < mVarNum );

  int vpos = 0;
  std::vector<SatLiteral> alits(na);
  for ( int i = 0; i < na; ++ i ) {
    alits[i] = mVarList[vpos];
    ++ vpos;
  }
  std::vector<SatLiteral> blits(nb);
  for ( int i = 0; i < nb; ++ i ) {
    blits[i] = mVarList[vpos];
    ++ vpos;
  }
  SatLiteral ilit = mVarList[vpos];
  ++ vpos;
  std::vector<SatLiteral> slits(ns);
  for ( int i = 0; i < ns; ++ i ) {
    slits[i] = mVarList[vpos];
    ++ vpos;
  }
  SatLiteral olit = mVarList[vpos];
  ++ vpos;

  mSolver.add_adder(alits, blits, ilit, slits, olit);

  int nexp = (1 << vpos);
  std::vector<int> vals(nexp);
  int amask = (1 << na) - 1;
  int ashift = 0;
  int bmask = (1 << nb) - 1;
  int bshift = na;
  int ishift = na + nb;
  int smask = (1 << ns) - 1;
  int sshift = na + nb + 1;
  int oshift = na + nb + 1 + ns;
  for ( int bit = 0; bit < nexp; ++ bit ) {
    int a = (bit >> ashift) & amask;
    int b = (bit >> bshift) & bmask;
    int i = (bit >> ishift) & 1;
    int s = (bit >> sshift) & smask;
    int o = (bit >> oshift) & 1;
    int s_exp = a + b + i;
    int o_exp = 0;
    if ( s_exp >= (1 << ns) ) {
      o_exp = 1;
      s_exp -= (1 << ns);
    }
    if ( s == s_exp && o == o_exp ) {
      vals[bit] = 1;
    }
    else {
      vals[bit] = 0;
    }
  }

  SatTestFixture::check(vpos, vals);
}

TEST_P(AdderTest, add_adder_4_4_4)
{
  check(4, 4, 4);
}

TEST_P(AdderTest, add_adder_4_3_4)
{
  check(4, 3, 4);
}

TEST_P(AdderTest, add_adder_4_2_4)
{
  check(4, 2, 4);
}

TEST_P(AdderTest, add_adder_3_4_4)
{
  check(3, 4, 4);
}

TEST_P(AdderTest, add_adder_2_4_4)
{
  check(2, 4, 4);
}

TEST_P(AdderTest, add_adder_3_3_4)
{
  check(3, 3, 4);
}

TEST_P(AdderTest, add_adder_3_2_4)
{
  check(3, 2, 4);
}

TEST_P(AdderTest, add_adder_3_1_4)
{
  check(3, 1, 4);
}

TEST_P(AdderTest, add_adder_3_1_5)
{
  check(3, 1, 5);
}

INSTANTIATE_TEST_SUITE_P(SatSolverTest,
			 AdderTest,
			 ::testing::Values("lingeling", "glueminisat2", "minisat2", "minisat",
					   "ymsat1", "ymsat2", "ymsat1_old"));

END_NAMESPACE_YM
